File: java/src/main/java/challenges/aoc2024/day3/service/ProgramService.java
package challenges.aoc2024.day3.service;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;

import challenges.aoc2024.day3.model.ComputerMemorySection;
import challenges.aoc2024.day3.model.ImmutableComputerMemorySection;
import challenges.aoc2024.day3.util.DayThreeConstants;

public class ProgramService {
    
    public ProgramService() {
        // Empty constructor for dependency injection
    }

    public Integer calculateTotalOfMultiplicationInstructions(String rawMemory, boolean conditionalsIntact) {

        ComputerMemorySection memory;
        
        if (conditionalsIntact) {
            final Matcher matcher = DayThreeConstants.DO_DONT_SECTION_PATTERN.matcher(rawMemory);

            final List<String> toCheck = new ArrayList<>(List.of(rawMemory.substring(0, rawMemory.indexOf("don't()"))));
            final String after = rawMemory.substring(rawMemory.lastIndexOf("do()") + 4);

            toCheck.addAll(DayThreeConstants.DO_DONT_SECTION_PATTERN.matcher(after).results()
                .map(match -> match.group(1))
                .toList());

            toCheck.addAll(matcher.results()
                .map(match -> match.group(1))
                .toList());

            memory = ImmutableComputerMemorySection.builder()
                .instructions(toCheck.stream()
                    .map(ComputerMemorySection::of)
                    .map(ComputerMemorySection::getInstructions)
                    .flatMap(List::stream)
                    .toList())
                .build();
        } else {
            memory = ComputerMemorySection.of(rawMemory);
        }

        return memory.getInstructions().stream()
            .map(instruction -> instruction.getFirstValue() * instruction.getSecondValue())
            .mapToInt(Integer::intValue)
            .sum();
    }
}

File: scala/src/main/scala/challenges/aoc2024/day3/service/ProgramService.scala
package challenges.aoc2024.day3.service

import challenges.aoc2024.day3.model.ComputerMemorySection
import challenges.aoc2024.day3.util.DayThreeConstants

import scala.util.matching.Regex

class ProgramService {

  private val doStatement: String = "do()"
  private val dontStatement: String = "don't()"

  def calculateTotalOfMultiplicationInstructions(rawMemory: String, conditionalsIntact: Boolean): Int = {

    val memory: ComputerMemorySection =
      if conditionalsIntact then
        val regex: Regex = DayThreeConstants.doDontSectionRegex
  
        var toCheck: List[String] = List(rawMemory.substring(0, rawMemory.indexOf(dontStatement)))
        val after: String = rawMemory.substring(rawMemory.lastIndexOf(doStatement) + 4)
  
        if !after.contains(dontStatement) then
          toCheck = toCheck :+ after
  
        toCheck = toCheck ++ regex.findAllMatchIn(rawMemory)
          .map(_.group(1))
  
        ComputerMemorySection(
          toCheck
            .map(ComputerMemorySection.of)
            .flatMap(_.instructions)
        )
  
      else
        ComputerMemorySection.of(rawMemory)

    memory.instructions
      .map(instruction => instruction.firstValue * instruction.secondValue)
      .sum
  }
}
